### ISSUE TITLE
Unable to build tf-master

### ISSUE THREAD

**Issue by iperov** [tensorflow/tensorflow#43664]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 7
- TensorFlow installed from (source or binary): source 
- TensorFlow version: master
- Python version: 3.6
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source): 3.5.0
- GCC/Compiler version (if compiling from source): MSVC 2017
- CUDA/cuDNN version: 11.1 / 7.6.0
- GPU model and memory: RTX 2080 TI

**Describe the problem**

unable to build 

**Any other info / logs**

```
[1,694 / 2,066] Compiling tensorflow/core/framework/graph.pb.cc; 2s local ... (4
 actions, 3 running)
ERROR: E:/dp/tools/tensorflow/tensorflow-master/tensorflow/core/platform/windows
/BUILD:151:11: C++ compilation of rule '//tensorflow/core/platform/windows:platf
orm_port' failed (Exit 2): python.exe failed: error executing command
  cd E:/dp/_internal/_e/up/_bazel_administrator/n2u4av42/execroot/org_tensorflow

  SET INCLUDE=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\T
ools\MSVC\14.16.27023\ATLMFC\include;C:\Program Files (x86)\Microsoft Visual Stu
dio\2017\Community\VC\Tools\MSVC\14.16.27023\include;C:\Program Files (x86)\Wind
ows Kits\NETFXSDK\4.6.1\include\um;C:\Program Files (x86)\Windows Kits\10\includ
e\10.0.17763.0\ucrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\
shared;C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um;C:\Program
 Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt;C:\Program Files (x86)\W
indows Kits\10\include\10.0.17763.0\cppwinrt
    SET LIB=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Too
ls\MSVC\14.16.27023\ATLMFC\lib\x64;C:\Program Files (x86)\Microsoft Visual Studi
o\2017\Community\VC\Tools\MSVC\14.16.27023\lib\x64;C:\Program Files (x86)\Window
s Kits\NETFXSDK\4.6.1\lib\um\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0
.17763.0\ucrt\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64
;
    SET PATH=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\To
ols\MSVC\14.16.27023\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Stu
dio\2017\Community\Common7\IDE\VC\VCPackages;C:\Program Files (x86)\Microsoft SD
Ks\TypeScript\3.1;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\
Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\Program Files (x86)\Microso
ft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFound
ation\Team Explorer;C:\Program Files (x86)\Microsoft Visual Studio\2017\Communit
y\MSBuild\15.0\bin\Roslyn;C:\Program Files (x86)\Microsoft Visual Studio\2017\Co
mmunity\Team Tools\Performance Tools\x64;C:\Program Files (x86)\Microsoft Visual
 Studio\2017\Community\Team Tools\Performance Tools;C:\Program Files (x86)\Micro
soft Visual Studio\Shared\Common\VSPerfCollectionTools\\x64;C:\Program Files (x8
6)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTools\;C:\Program Files
 (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.6.1 Tools\x64\;C:\Program Files
 (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Micro
soft\FSharp\;C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64;C:\Prog
ram Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Microsoft Visual
Studio\2017\Community\\MSBuild\15.0\bin;C:\Windows\Microsoft.NET\Framework64\v4.
0.30319;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\ID
E\;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\;
;C:\Windows\system32;C:\Program Files (x86)\Microsoft Visual Studio\2017\Communi
ty\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin;C:\Program Files (x86)
\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\C
Make\Ninja
    SET PWD=/proc/self/cwd
    SET RUNFILES_MANIFEST_ONLY=1
    SET TEMP=E:\DP\_internal\_e\_t
    SET TMP=E:\DP\_internal\_e\_t
  E:/DP/python/3.6.8/python.exe -B external/local_config_cuda/crosstool/windows/
msvc_wrapper_for_nvcc.py /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x060
0 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /D_SILENCE_STDEXT_HASH_DE
PRECATION_WARNINGS /bigobj /Zm500 /J /Gy /GF /EHsc /wd4351 /wd4291 /wd4250 /wd49
96 /I. /Ibazel-out/x64_windows-opt-exec-50AE0418/bin /Iexternal/eigen_archive /I
bazel-out/x64_windows-opt-exec-50AE0418/bin/external/eigen_archive /Iexternal/co
m_google_absl /Ibazel-out/x64_windows-opt-exec-50AE0418/bin/external/com_google_
absl /Iexternal/snappy /Ibazel-out/x64_windows-opt-exec-50AE0418/bin/external/sn
appy /Iexternal/eigen_archive /Ibazel-out/x64_windows-opt-exec-50AE0418/bin/exte
rnal/eigen_archive /DEIGEN_MPL2_ONLY /DEIGEN_MAX_ALIGN_BYTES=64 /DEIGEN_HAS_TYPE
_TRAITS=0 /D__CLANG_SUPPORT_DYN_ANNOTATION__ /DTF_USE_SNAPPY /showIncludes /MD /
O2 /DNDEBUG /D_USE_MATH_DEFINES -DWIN32_LEAN_AND_MEAN -DNOGDI /experimental:prep
rocessor /std:c++14 -DGOOGLE_CUDA=1 -DTENSORFLOW_USE_NVCC=1 -DTENSORFLOW_MONOLIT
HIC_BUILD /DPLATFORM_WINDOWS /DEIGEN_HAS_C99_MATH /DTENSORFLOW_USE_EIGEN_THREADP
OOL /DEIGEN_AVOID_STL_ARRAY /Iexternal/gemmlowp /wd4018 /wd4577 /DNOGDI /DTF_COM
PILE_LIBRARY /Fobazel-out/x64_windows-opt-exec-50AE0418/bin/tensorflow/core/plat
form/windows/_objs/platform_port/port.obj /c tensorflow/core/platform/windows/po
rt.cc
Execution platform: @local_execution_config_platform//:platform
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\stdio.h(378): w
arning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\stdio.h(2437):
warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\corecrt_search.
h(188): warning C5105: macro expansion producing 'defined' has undefined behavio
r
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\stdlib.h(79): w
arning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\stdlib.h(1286):
 warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\corecrt_memory.
h(76): warning C5105: macro expansion producing 'defined' has undefined behavior

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\corecrt_wstring
.h(573): warning C5105: macro expansion producing 'defined' has undefined behavi
or
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\string.h(531):
warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\corecrt_math.h(
44): warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\corecrt_math.h(
963): warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\malloc.h(173):
warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\float.h(328): w
arning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\sys/stat.h(87):
 warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\sys/stat.h(120)
: warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\sys/stat.h(217)
: warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\ctype.h(241): w
arning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\winbase.h(9254):
warning C5105: macro expansion producing 'defined' has undefined behavior
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(487): war
ning C5103: pasting '/' and '/' does not result in a valid preprocessing token
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared\wtypes.h(745)
: note: in expansion of macro '_VARIANT_BOOL'

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(487): err
or C2059: syntax error: '/'
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(487): err
or C2238: unexpected token(s) preceding ';'

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(502): war
ning C5103: pasting '/' and '/' does not result in a valid preprocessing token
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared\wtypes.h(745)
: note: in expansion of macro '_VARIANT_BOOL'

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(502): err
or C2059: syntax error: '/'
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\oaidl.h(502): err
or C2238: unexpected token(s) preceding ';'

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\propidlbase.h(319
): warning C5103: pasting '/' and '/' does not result in a valid preprocessing t
oken
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared\wtypes.h(745)
: note: in expansion of macro '_VARIANT_BOOL'

C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\propidlbase.h(319
): error C2059: syntax error: '/'
C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um\propidlbase.h(319
): error C2238: unexpected token(s) preceding ';'

Target //tensorflow/tools/pip_package:build_pip_package failed to build
INFO: Elapsed time: 481.272s, Critical Path: 17.16s
INFO: 458 processes: 458 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
The system cannot find the path specified.
```

seems like _VARIANT_BOOL is replaced thus generates error in windows kits library.

**Comment by Saduf2019** [comment]
@iperov 
Please share the steps followed before you encounter this issue, please follow [this guide ](https://www.tensorflow.org/install/source_windows) and let us know.

**Comment by gloriouskilka** [comment]
Solution: update your Windows SDK to 10.0.19041.0, because there is a bug in 10.0.17763.0 version headers

**Comment by YoannNicod** [comment]
I can't understand how @iperov could have closed this issue months before someone gave the solution.
This shows that tf dev team does not care at all abou tf C++ usage, and even less about tf used on Windows.
Because yeah, I want to use tf in C++ on Windows, how crazy am I, right ??

Thank you @gloriouskilka , your answer is perfectly correct and helped me resolve the issue I also had.
Now I need to face the 60 000 DLL symbols limitation on Windows.

I take the opportunity to say that the tf compiling process is a shame for a framework this big and serious. If everybody made compiling that hard for their projects, the world would definitely be a worse place.

**Comment by iperov** [comment]
I am migrating to pytorch.
TF is dead and it was a wrong choice.

**Comment by gloriouskilka** [comment]
@YoannNicod You're welcome!
Partially agree, I'd like to rephrase, if you please: if everyone would close issues without providing the solution... :)

Speaking of this particular problem, there should be pretty neat Windows CI build check infrastructure to detect such weird compiling error (What? Windows SDK version related bug? Seriously?), I doubt that could be predicted or/and automatically checked.

**Comment by YoannNicod** [comment]
@gloriouskilka 
I am not talking _only_ about this issue, but about the entire tensorflow compiling process which is really a pain in the ***.
Though I agree with you, Win SDK related bug is unexpected, it shows that tf is almost never compiled on windows for C++.

By the way, I gave up compiling tensorflow, the generated dll doesn't expose most of the usefull symbols (because of the 64k symbols dll windows limitation, apparently). I am NOT going to fix tensorflow code myself as suggested in a 3 years old tf issue (https://github.com/tensorflow/tensorflow/issues/23542).

I have defended Tensorflow long enough, I think it is time that I too migrate to PyTorch.

**Comment by iperov** [comment]
@YoannNicod 
there are a lot of problems with basic usage of TF, not only with compilation. 
I wrote several issues , but nothing were fixed.
Especially with RTX 3x. 
I had huge headache when RTX 3k was released and DeepFaceLab had to be upgraded to work with it.
My new project will work on amazing pytorch, that just works.

I think the biggest mistake of google, was to start upgrade tf to pytorch-like eager-execution, which is produced tons of bugs and anal pain of developers. Graph system was fine and better optimized.

press F to TF.
